logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb

```diff
@@ -190,28 +192,29 @@ def initialize(queue, size, wait)
         @size = size
         @wait = wait
 
-        @originals = Hash.new
-
         # TODO: disabled for https://github.com/elastic/logstash/issues/6055 - will have to properly refactor
         # @cancelled = Hash.new
 
+        @originals = java.util.HashSet.new size, 1
         @generated = Hash.new
         @iterating_temp = Hash.new
         @iterating = false # Atomic Boolean maybe? Although batches are not shared across threads
         @acked_batch = nil
       end
 
       def read_next
-        @size.times do |t|
-          event = @queue.poll(@wait)
-          return if event.nil? # queue poll timed out
-
-          @originals[event] = true
+        @queue.queue.drainTo(@originals, @size)
+        if @originals.size == 0 
+          e = @queue.poll(@wait)
```

Comment:
@andrewvc we only `poll` if `drainTo` didn't give us any events. We still have the same guarantee that a batch will contain at least one event (actually it's more likely now since we're making the `drainTo` call as well. We don't have the timed `poll` for more than one event and are breaking guarantees there yes, but:
If you think about it, practically we have a much stronger guarantee now. We pretty much have a `100%` chance to get a completely filled batch in the first call to `@queue.queue.drainTo(@originals, @size)` anyway. 
Now that the `Queue` has capacity, it doesn't really make sense to have a timeout here since load spikes are handled by the `Queue`'s capacity + fast producers guarantee 100% full batches. Slow producers may see smaller average batch fill sizes, but:
1. This is a very rare case afaik? Filters and outputs are the bottlenecks in all example I benchmarked so far. 
2. Obviously, you could have a slow Kafka input or so but even then the input is mostly kinda "bursty" and should see larger batch sizes from not blocking on the push to the Queue all that much.

Hope that makes sense :)

PS: IMO, we shouldn't even have a `@wait` there and just poll indefinitely to avoid zero length buckets, but